File: crates/leveler-remote-protocol/src/envelope.rs
//! The end-to-end signed envelope: what a relay is structurally unable to forge.
//!
//! Every payload that a paired device and a runtime exchange through a relay
//! travels inside a [`SignedEnvelope`]. The relay routes it and cannot alter it:
//! it holds neither key, and the signature covers the routing-relevant header
//! fields, not just the body. That is the whole point — the design's threat
//! model assumes a compromised relay, so "the relay said so" is never evidence.
//!
//! ## What the signature covers, and why each field is in it
//!
//! The canonical string is the exact byte sequence that gets signed:
//!
//! ```text
//! {v}|{sender}|{sender_id}|{recipient_id}|{stream_id}|{seq}|{ts}|{content_type}|{digest_hex}
//! ```
//!
//! - `recipient_id` binds the envelope to **one** peer. Without it a device
//!   paired to two hosts could have a frame it signed for host A replayed by a
//!   malicious relay into host B, which would verify it happily. Verification
//!   therefore rejects any envelope not addressed to the verifier.
//! - `stream_id` and `seq` bind it to a position in a stream, so a frame cannot
//!   be lifted into a different conversation or replayed within one.
//! - `digest_hex` is `SHA-256(payload_raw)` — the raw bytes, *not* the base64
//!   text. Hashing the encoding instead would let two different encodings of
//!   one payload share a signature.
//!
//! Ids are restricted to [`ID_CHARSET_DESCRIPTION`] before they ever reach the
//! canonical string. `|` is the field separator, so an id allowed to contain it
//! could shift the field boundaries and make two different envelopes produce
//! one identical signing input.

use base64::Engine as _;
use serde::{Deserialize, Serialize};

use crate::keys::{SigningKey, VerifyingKey};

/// What a verifier knows about itself and its peer.
///
/// `expected_recipient_id` is the verifier's own id — not a value taken from
/// the envelope. Reading it from the frame would defeat the audience binding
/// entirely.
#[derive(Debug, Clone, Copy)]
pub struct VerifyParams<'a> {
    pub expected_recipient_id: &'a str,
    /// The peer's key from a trusted local source: the agent's TOFU
    /// `devices.json`, or the `runtime_pubkey` the APP anchored from the QR.
    /// Never a key the relay supplied alongside the frame.
    pub public_key: &'a VerifyingKey,
    /// The verifier's clock, in the envelope timestamp format.
    pub now: &'a str,
}

/// Human-readable form of the id rule, for error messages and docs.
pub const ID_CHARSET_DESCRIPTION: &str = "1-64 chars from [A-Za-z0-9_.:-]";

/// Maximum accepted clock skew, in seconds, between an envelope's `ts` and the
/// verifier's clock. Both directions.
pub const TIMESTAMP_WINDOW_SECS: i64 = 120;

/// Which side signed the envelope.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Sender {
    Device,
    Runtime,
}

impl Sender {
    /// The literal used in the canonical string.
    pub const fn as_str(self) -> &'static str {
        match self {
            Sender::Device => "device",
            Sender::Runtime => "runtime",
        }
    }
}

/// What the payload is. Signed, so a relay cannot re-label a frame.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ContentType {
    SessionUpstream,
    SessionDownstream,
    RpcRequest,
    RpcResponse,
}

impl ContentType {
    pub const fn as_str(self) -> &'static str {
        match self {
            ContentType::SessionUpstream => "session_upstream",
            ContentType::SessionDownstream => "session_downstream",
            ContentType::RpcRequest => "rpc_request",
            ContentType::RpcResponse => "rpc_response",
        }
    }
}

/// Why an envelope was refused. These map to the design's error-code catalogue;
/// [`EnvelopeError::code`] is the wire spelling.
#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
pub enum EnvelopeError {
    #[error("invalid {field}: expected {ID_CHARSET_DESCRIPTION}")]
    InvalidId { field: &'static str },
    #[error("invalid timestamp: expected YYYY-MM-DDTHH:MM:SSZ")]
    InvalidTimestamp,
    #[error("invalid base64 in {field}")]
    InvalidBase64 { field: &'static str },
    #[error("unsupported envelope version: {got}")]
    UnsupportedVersion { got: u8 },
    #[error("envelope addressed to {addressed_to}, not {expected}")]
    RecipientMismatch {
        expected: String,
        addressed_to: String,
    },
    #[error("timestamp outside the ±{TIMESTAMP_WINDOW_SECS}s window")]
    TimestampOutOfWindow,
    #[error("signature does not verify")]
    SignatureInvalid,
    #[error("invalid Ed25519 key material")]
    InvalidKey,
}

impl EnvelopeError {
    /// The wire error code from the design's catalogue.
    pub fn code(&self) -> &'static str {
        match self {
            EnvelopeError::InvalidId { .. }
            | EnvelopeError::InvalidTimestamp
            | EnvelopeError::InvalidBase64 { .. }
            | EnvelopeError::UnsupportedVersion { .. }
            | EnvelopeError::InvalidKey => "invalid_frame",
            EnvelopeError::RecipientMismatch { .. } => "recipient_mismatch",
            EnvelopeError::TimestampOutOfWindow => "replay",
            EnvelopeError::SignatureInvalid => "signature_invalid",
        }
    }
}

/// The only envelope version this build speaks.
pub const ENVELOPE_VERSION: u8 = 1;

/// A signed, relay-opaque frame. Field order here is also the JSON order.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct SignedEnvelope {
    pub v: u8,
    pub sender: Sender,
    pub sender_id: String,
    /// The peer this envelope is for. Verification fails unless it names the
    /// verifier — this is what stops cross-runtime replay by a relay.
    pub recipient_id: String,
    /// The session stream, or `rpc:{uuid}` for a request/response pair.
    pub stream_id: String,
    pub seq: u64,
    /// UTC RFC3339, seconds precision, `Z`-suffixed, no fractional part.
    pub ts: String,
    pub content_type: ContentType,
    /// Standard base64 (with padding) of the raw payload bytes.
    pub payload_b64: String,
    /// Standard base64 of the 64-byte Ed25519 signature.
    pub sig_b64: String,
}

/// Whether `id` may appear in a canonical string.
///
/// Rejecting `|` is the load-bearing part: it is the field separator, so an id
/// containing one could make two distinct envelopes canonicalize identically.
/// The rest of the restriction keeps ids to a shape every client can round-trip
/// without escaping questions.
pub fn id_is_valid(id: &str) -> bool {
    !id.is_empty()
        && id.len() <= 64
        && id
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || matches!(b, b'_' | b'.' | b':' | b'-'))
}

impl SignedEnvelope {
    /// Rebuild the exact byte sequence that the signature covers.
    ///
    /// Fails if any id would corrupt the field structure, or if `ts` is not in
    /// the one accepted timestamp format — both are checked here so that no
    /// caller can sign or verify a string it did not fully constrain.
    pub fn canonical_string(&self) -> Result<String, EnvelopeError> {
        if self.v != ENVELOPE_VERSION {
            return Err(EnvelopeError::UnsupportedVersion { got: self.v });
        }
        for (field, value) in [
            ("sender_id", &self.sender_id),
            ("recipient_id", &self.recipient_id),
            ("stream_id", &self.stream_id),
        ] {
            if !id_is_valid(value) {
                return Err(EnvelopeError::InvalidId { field });
            }
        }
        parse_timestamp(&self.ts)?;

        let payload = self.payload()?;
        let digest = ring::digest::digest(&ring::digest::SHA256, &payload);
        let digest_hex = hex_lower(digest.as_ref());

        Ok(format!(
            "{}|{}|{}|{}|{}|{}|{}|{}|{}",
            self.v,
            self.sender.as_str(),
            self.sender_id,
            self.recipient_id,
            self.stream_id,
            self.seq,
            self.ts,
            self.content_type.as_str(),
            digest_hex,
        ))
    }

    /// Build and sign an envelope over `payload_raw`.
    ///
    /// Refuses to produce an envelope whose ids or timestamp would not
    /// canonicalize, so a malformed frame never reaches the wire.
    #[allow(clippy::too_many_arguments)]
    pub fn sign(
        key: &SigningKey,
        sender: Sender,
        sender_id: &str,
        recipient_id: &str,
        stream_id: &str,
        seq: u64,
        ts: &str,
        content_type: ContentType,
        payload_raw: &[u8],
    ) -> Result<Self, EnvelopeError> {
        let mut envelope = Self {
            v: ENVELOPE_VERSION,
            sender,
            sender_id: sender_id.to_string(),
            recipient_id: recipient_id.to_string(),
            stream_id: stream_id.to_string(),
            seq,
            ts: ts.to_string(),
            content_type,
            payload_b64: base64::engine::general_purpose::STANDARD.encode(payload_raw),
            sig_b64: String::new(),
        };

        // Canonicalizing first is what enforces the id and timestamp rules on
        // the signing path: an envelope that cannot canonicalize is never signed.
        let canonical = envelope.canonical_string()?;
        let signature = key.sign_bytes(canonical.as_bytes());
        envelope.sig_b64 = base64::engine::general_purpose::STANDARD.encode(signature);
        Ok(envelope)
    }

    /// Check the envelope against `params` and return the payload it carries.
    ///
    /// Order matters: structure, then audience, then freshness, then the
    /// signature. A caller must treat the payload as untrusted until this
    /// returns `Ok`.
    pub fn verify(&self, params: &VerifyParams<'_>) -> Result<Vec<u8>, EnvelopeError> {
        let canonical = self.canonical_string()?;

        if self.recipient_id != params.expected_recipient_id {
            return Err(EnvelopeError::RecipientMismatch {
                expected: params.expected_recipient_id.to_string(),
                addressed_to: self.recipient_id.clone(),
            });
        }

        let sent_at = parse_timestamp(&self.ts)?;
        let now = parse_timestamp(params.now)?;
        if (now - sent_at).num_seconds().abs() > TIMESTAMP_WINDOW_SECS {
            return Err(EnvelopeError::TimestampOutOfWindow);
        }

        let signature = base64::engine::general_purpose::STANDARD
            .decode(&self.sig_b64)
            .map_err(|_| EnvelopeError::InvalidBase64 { field: "sig_b64" })?;
        if !params
            .public_key
            .verify_bytes(canonical.as_bytes(), &signature)
        {
            return Err(EnvelopeError::SignatureInvalid);
        }

        self.payload()
    }

    /// The raw payload bytes this envelope carries.
    pub fn payload(&self) -> Result<Vec<u8>, EnvelopeError> {
        base64::engine::general_purpose::STANDARD
            .decode(&self.payload_b64)
            .map_err(|_| EnvelopeError::InvalidBase64 {
                field: "payload_b64",
            })
    }
}

/// Accepts exactly `YYYY-MM-DDTHH:MM:SSZ`, returning the instant it names.
///
/// Deliberately strict. A looser parser would accept offsets or fractional
/// seconds, and two peers disagreeing on how to render one instant would
/// disagree on the canonical string — an interoperability failure that presents
/// as an invalid signature.
fn parse_timestamp(ts: &str) -> Result<chrono::DateTime<chrono::Utc>, EnvelopeError> {
    use chrono::TimeZone as _;

    let naive = chrono::NaiveDateTime::parse_from_str(ts, "%Y-%m-%dT%H:%M:%SZ")
        .map_err(|_| EnvelopeError::InvalidTimestamp)?;
    chrono::Utc
        .from_local_datetime(&naive)
        .single()
        .ok_or(EnvelopeError::InvalidTimestamp)
}

pub(crate) fn hex_lower(bytes: &[u8]) -> String {
    use std::fmt::Write as _;

    bytes.iter().fold(String::new(), |mut out, byte| {
        let _ = write!(out, "{byte:02x}");
        out
    })
}

File: apps/leveler-mobile/lib/protocol/envelope.dart
/// The signed envelope, as the phone must build and check it.
///
/// This file is the app's half of a two-language agreement. The host's Rust
/// implementation and this one must produce byte-identical canonical strings,
/// or every signature disagrees and nothing works — so the rules are restated
/// here rather than paraphrased, and `test/envelope_golden_test.dart` replays
/// the same answer key the Rust tests use.
///
/// What the phone must never do is act on an unverified frame. A relay carries
/// these envelopes and is assumed hostile: it can drop, delay, reorder and
/// duplicate them, and the only reason it cannot *author* one is that the
/// signature covers the header as well as the body.
library;

import 'dart:convert';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';

import 'ids.dart';

/// Seconds a frame's timestamp may differ from the receiver's clock.
const int timestampWindowSeconds = 120;

/// The envelope version this build speaks.
const String envelopeVersion = '1';

enum FrameSender {
  device,
  runtime;

  String get wire => name;

  static FrameSender? fromWire(String value) {
    for (final sender in FrameSender.values) {
      if (sender.wire == value) return sender;
    }
    return null;
  }
}

enum ContentType {
  sessionUpstream('session_upstream'),
  sessionDownstream('session_downstream'),
  rpcRequest('rpc_request'),
  rpcResponse('rpc_response');

  const ContentType(this.wire);
  final String wire;

  static ContentType? fromWire(String value) {
    for (final type in ContentType.values) {
      if (type.wire == value) return type;
    }
    return null;
  }
}

/// Why a frame was refused. The codes match the host's catalogue so the two
/// sides can be read together in a bug report.
enum EnvelopeError {
  invalidFrame('invalid_frame'),
  recipientMismatch('recipient_mismatch'),
  signatureInvalid('signature_invalid'),
  staleTimestamp('stale_timestamp');

  const EnvelopeError(this.code);
  final String code;
}

class EnvelopeException implements Exception {
  EnvelopeException(this.error, [this.detail]);
  final EnvelopeError error;
  final String? detail;

  String get code => error.code;

  @override
  String toString() => 'EnvelopeException(${error.code}${detail == null ? '' : ': $detail'})';
}

/// One envelope, exactly as it travels.
class SignedEnvelope {
  SignedEnvelope({
    required this.version,
    required this.sender,
    required this.senderId,
    required this.recipientId,
    required this.streamId,
    required this.seq,
    required this.ts,
    required this.contentType,
    required this.payloadB64,
    required this.sigB64,
  });

  final String version;
  final FrameSender sender;
  final String senderId;
  final String recipientId;
  final String streamId;
  final int seq;
  final String ts;
  final ContentType contentType;
  final String payloadB64;
  final String sigB64;

  factory SignedEnvelope.fromJson(Map<String, dynamic> json) {
    final sender = FrameSender.fromWire(json['sender'] as String? ?? '');
    final contentType = ContentType.fromWire(json['content_type'] as String? ?? '');
    final seq = json['seq'];
    if (sender == null || contentType == null || seq is! int) {
      throw EnvelopeException(EnvelopeError.invalidFrame, 'unrecognised envelope fields');
    }
    return SignedEnvelope(
      // Serialised as a number by the host; compared as its decimal text.
      version: '${json['v']}',
      sender: sender,
      senderId: json['sender_id'] as String? ?? '',
      recipientId: json['recipient_id'] as String? ?? '',
      streamId: json['stream_id'] as String? ?? '',
      seq: seq,
      ts: json['ts'] as String? ?? '',
      contentType: contentType,
      payloadB64: json['payload_b64'] as String? ?? '',
      sigB64: json['sig_b64'] as String? ?? '',
    );
  }

  Map<String, dynamic> toJson() => {
        'v': int.parse(version),
        'sender': sender.wire,
        'sender_id': senderId,
        'recipient_id': recipientId,
        'stream_id': streamId,
        'seq': seq,
        'ts': ts,
        'content_type': contentType.wire,
        'payload_b64': payloadB64,
        'sig_b64': sigB64,
      };

  /// The raw payload bytes. Standard base64, with padding.
  Uint8List payloadBytes() {
    try {
      return Uint8List.fromList(base64.decode(payloadB64));
    } on FormatException {
      throw EnvelopeException(EnvelopeError.invalidFrame, 'payload is not base64');
    }
  }

  /// The payload decoded as JSON, once the signature has been checked.
  Map<String, dynamic> payloadJson() {
    final decoded = jsonDecode(utf8.decode(payloadBytes()));
    if (decoded is! Map<String, dynamic>) {
      throw EnvelopeException(EnvelopeError.invalidFrame, 'payload is not a JSON object');
    }
    return decoded;
  }
}

/// Build the exact string a signature covers.
///
/// Every id is checked first: a separator inside one would shift the field
/// boundaries, which is precisely how two different frames could be made to
/// share a signature.
Future<String> canonicalString(SignedEnvelope envelope) async {
  for (final id in [envelope.senderId, envelope.recipientId, envelope.streamId]) {
    if (!isValidId(id)) {
      throw EnvelopeException(EnvelopeError.invalidFrame, 'illegal id: $id');
    }
  }
  if (parseTimestamp(envelope.ts) == null) {
    throw EnvelopeException(EnvelopeError.invalidFrame, 'timestamp must be YYYY-MM-DDTHH:MM:SSZ');
  }
  if (envelope.seq < 0) {
    throw EnvelopeException(EnvelopeError.invalidFrame, 'seq must not be negative');
  }

  // The digest is over the *raw payload bytes*, not over their base64 text.
  final digest = await Sha256().hash(envelope.payloadBytes());
  final digestHex = digest.bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();

  return [
    envelope.version,
    envelope.sender.wire,
    envelope.senderId,
    envelope.recipientId,
    envelope.streamId,
    '${envelope.seq}',
    envelope.ts,
    envelope.contentType.wire,
    digestHex,
  ].join('|');
}

/// Parse the one accepted timestamp shape, or null.
///
/// Deliberately strict: accepting other ISO-8601 forms would let two peers
/// disagree about the bytes being signed while both believing they parsed the
/// same instant.
DateTime? parseTimestamp(String ts) {
  final pattern = RegExp(r'^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})Z$');
  final match = pattern.firstMatch(ts);
  if (match == null) return null;
  try {
    return DateTime.utc(
      int.parse(match.group(1)!),
      int.parse(match.group(2)!),
      int.parse(match.group(3)!),
      int.parse(match.group(4)!),
      int.parse(match.group(5)!),
      int.parse(match.group(6)!),
    );
  } on ArgumentError {
    return null;
  }
}

/// Format an instant the way every signer must.
String formatTimestamp(DateTime instant) {
  final utc = instant.toUtc();
  String two(int value) => value.toString().padLeft(2, '0');
  return '${utc.year.toString().padLeft(4, '0')}-${two(utc.month)}-${two(utc.day)}'
      'T${two(utc.hour)}:${two(utc.minute)}:${two(utc.second)}Z';
}

/// Verify a frame and return its payload bytes.
///
/// Throws rather than returning null: a caller that forgot to check a boolean
/// would be acting on relay-controlled data, and that is the one mistake this
/// whole design exists to prevent.
Future<Uint8List> verifyEnvelope(
  SignedEnvelope envelope, {
  required String expectedRecipientId,
  required SimplePublicKey publicKey,
  required DateTime now,
}) async {
  // Audience first: a genuine frame addressed to another host is still not for
  // us, and checking it before the signature keeps the cheap rejection cheap.
  if (envelope.recipientId != expectedRecipientId) {
    throw EnvelopeException(EnvelopeError.recipientMismatch,
        'addressed to ${envelope.recipientId}');
  }

  final canonical = await canonicalString(envelope);

  final Uint8List signature;
  try {
    signature = Uint8List.fromList(base64.decode(envelope.sigB64));
  } on FormatException {
    throw EnvelopeException(EnvelopeError.invalidFrame, 'signature is not base64');
  }

  final verified = await Ed25519().verify(
    utf8.encode(canonical),
    signature: Signature(signature, publicKey: publicKey),
  );
  if (!verified) {
    throw EnvelopeException(EnvelopeError.signatureInvalid);
  }

  // Freshness last, so a stale frame is still reported as stale rather than as
  // a signature problem — the two mean very different things to a user.
  final stamped = parseTimestamp(envelope.ts)!;
  final skew = now.toUtc().difference(stamped).inSeconds.abs();
  if (skew > timestampWindowSeconds) {
    throw EnvelopeException(EnvelopeError.staleTimestamp, '${skew}s from this clock');
  }

  return envelope.payloadBytes();
}

/// Sign one payload as this device.
Future<SignedEnvelope> signEnvelope({
  required SimpleKeyPair keyPair,
  required String senderId,
  required String recipientId,
  required String streamId,
  required int seq,
  required DateTime now,
  required ContentType contentType,
  required List<int> payload,
}) async {
  final envelope = SignedEnvelope(
    version: envelopeVersion,
    sender: FrameSender.device,
    senderId: senderId,
    recipientId: recipientId,
    streamId: streamId,
    seq: seq,
    ts: formatTimestamp(now),
    contentType: contentType,
    payloadB64: base64.encode(payload),
    sigB64: '',
  );
  final canonical = await canonicalString(envelope);
  final signature = await Ed25519().sign(utf8.encode(canonical), keyPair: keyPair);
  return SignedEnvelope(
    version: envelope.version,
    sender: envelope.sender,
    senderId: envelope.senderId,
    recipientId: envelope.recipientId,
    streamId: envelope.streamId,
    seq: envelope.seq,
    ts: envelope.ts,
    contentType: envelope.contentType,
    payloadB64: envelope.payloadB64,
    sigB64: base64.encode(signature.bytes),
  );
}
